:sparkles: CombinedStatus Calculation (part 2)

vMaroon opened the pull request:
## Summary
This PR outlines the design for the implementation of the calculation of CombinedStatus objects.
The PR introduces new components that maintain the resolutions of combined-statuses, by maintaining a resolution entry per each of the latter (in the new CombinedStatusResolver), in which information from StatusCollectors and the evaluation of WorkStatuses are kept up to date.

The code is incomplete, and has TODOs that should cover finishing the maintenance of CombinedStatus objects without error propagation.

Note that the outlined design attempts to find balance between over-optimizing (in my opinion there is none), and the efficiency required due to the highly distributed nature of the feature while respecting eventual consistency and multiple information update sources. 

## Related issue(s)
- #2141

MikeSpreitzer commented on pkg/abstract/slices.go:
```diff
@@ -75,6 +75,19 @@ func SliceMap[Domain, Range any](slice []Domain, fn func(Domain) Range) []Range
 	return ans
 }
 
+func SliceToMap[Domain, Range any, Key comparable](slice []Domain, keyFn func(Domain) Key,
```
Perhaps name this `SliceToPrimitiveMap`, for consistency with the other names and anticipating the day when we have a general `Map` abstraction or simply not presuming that built-in maps are the only kind of map.

MikeSpreitzer commented on pkg/binding/bindingpolicy-resolution-broker.go:
```diff
@@ -44,6 +45,7 @@ type ResolutionBroker interface {
 // Resolution is a struct that represents the resolution of a binding policy.
 // It contains the destinations and object data for the resolution.
 type Resolution struct {
+	Name string
```
Note to self: why is the name needed here, doesn't the user already know the name?

MikeSpreitzer commented on pkg/status/cel-evaluator.go:
```diff
@@ -0,0 +1,111 @@
+// celEvaluator is a struct that holds the CEL environment
+// and provides a method to evaluate an expression with an unstructured object
+// as the context.
+type celEvaluator struct {
```
You saw https://github.com/kubestellar/kubestellar/blame/v0.23.0-alpha.3/api/control/v1alpha1/types.go#L416-L418 , right? No need to start from scratch.

MikeSpreitzer replied:
My point is that Kubernetes already has code that supports using CEL with Kubernetes API objects, let's use it.

There are a lot of particulars, regarding the particular uses of CEL in Kubernetes, do not get distracted by that. But, if you are interested in the background, you might start by looking at how the type-checking is done for the CEL expressions that appear in `x-kubernetes-validations` items in the OpenAPI schema in a CustomResourceDefinition. https://github.com/kubernetes/apiextensions-apiserver/blob/v0.28.2/pkg/apis/apiextensions/validation/validation.go#L73 validates a CustomResourceDefinnition. That calls `validateCustomResourceDefinitionSpec`, which calls `validateCustomResourceDefinitionValidation` (to validate what the CRD says about how to validate the objects whose Kind that CRD is defining), which calls `ValidateCustomResourceDefinitionOpenAPISchema`, which calls a Compile method at https://github.com/kubernetes/apiextensions-apiserver/blob/v0.28.2/pkg/apis/apiextensions/validation/validation.go#L1112, which parses and type-checks (ultimately using cel-go's Env.Compile) all the CEL expressions in the OpenAPI schema. But let me emphasize again, this is more meta and complex than the problem that we have.

https://github.com/kubernetes/apiserver/blob/v0.28.2/pkg/cel/common/schemas.go#L40 is the heart of the generic part. This func takes an OpenAPI schema (which we can expect the resources processed by this feature to have) and produces a DeclType, which is the thing used in type-checking. The status controller could use this to type-check each CEL expression against each resource (kind of object) that it gets applied to. The parsing and type-checking would only have to be done once per (resource, CEL expression), not repeated for every version of every individual workload object's status.

vMaroon replied:
Will be addressed in followup PR - logged in issue #2141.

MikeSpreitzer commented on pkg/status/cel-evaluator.go:
```diff
@@ -0,0 +1,111 @@
+
+// preprocessExpression replaces occurrences of $ with obj.
+func (e *celEvaluator) preprocessExpression(expression string) string {
+	return strings.ReplaceAll(expression, "$", "obj")
```
Hmm, did we never write down concrete `Expression` values?

Certainly a string-level replacement of dollar-sign with "obj" can not be valid, we should allow expressions like `something == "$"`, for example.

I was supposing that that accessing a field like `spec.replicas` would be done with an `Expression` like `spec.replicas`, for simplicity and conciseness. We _could_ instead define one CEL "variable" that holds the whole object. It is worth writing the decision down clearly somewhere prominent.

MikeSpreitzer commented on pkg/status/cel-evaluator.go:
```diff
@@ -0,0 +1,111 @@
+
+// Evaluate takes an expression and a Kubernetes raw object, and returns the
+// evaluation of the expression with the object as the context.
+func (e *celEvaluator) Evaluate(expression string, rawObj *runtime.RawExtension) (ref.Val, error) {
```
Ouch, reflection. Instant performance loser. Hmm, seems to be inherent in cel-go?

vMaroon replied:
Yep.

MikeSpreitzer commented on pkg/status/combinedstatus-resolution.go:
```diff
@@ -0,0 +1,308 @@
+)
+
+// combinedStatusResolution is a struct that represents the resolution of a
+// combinedstatus. A combinedstatus resolution maps to (binding, object).
```
It is not a general map, it is an injection AKA 1:1 mapping.

MikeSpreitzer commented on pkg/status/combinedstatus-resolution.go:
```diff
@@ -0,0 +1,308 @@
+//     calculating combinedFields efficient if present.
+//
+// - Ownership:
+//   - A combinedstatus resolution is owned by the binding that it is
```
"owned by" is a bit vague, can you say what is meant by that here?

vMaroon replied:
removed

MikeSpreitzer commented on pkg/status/combinedstatus-resolver.go:
```diff
@@ -0,0 +1,307 @@
+	statusCollectorNameToClauses map[string]statusCollectorClauses
+}
+
+type statusCollectorClauses struct {
```
This type looks like it has the same content as `v1alpha1.StatusCollectorSpec`, why does it exist?

vMaroon replied:
removed

MikeSpreitzer commented on pkg/status/combinedstatus-resolution.go:
```diff
@@ -0,0 +1,308 @@
+	// apart from the namespace, which is set to the associated clusterName.
+	// If a workstatus reference is mapped to nil, it means that the workstatus
+	// has not been evaluated against the statuscollector's clauses.
+	workStatusRefToData map[util.ObjectIdentifier]workStatusData
```
Why have entries in this map whose `filterEval` is `false`?

MikeSpreitzer replied:
Each statusCollectorData is specific to one workload object. All that is needed in the domain of this map is the WEC name (equivalently, mailbox namespace name).

MikeSpreitzer commented on pkg/status/combinedstatus-resolution.go:
```diff
@@ -0,0 +1,308 @@
+	filterEval bool
+
+	// groupByEval is a map of groupBy expression names to their evaluated values.
+	groupByEval map[string]any
```
Note to self: `any` here, `ref.Val` coming from cel-go's Eval.

MikeSpreitzer replied:
"Note to self" means only that, something for me to not forget; it is not an implied criticism.

MikeSpreitzer replied:
Another possible choice is the Value type from the API.

vMaroon replied:
Will be addressed in followup PR if sensible once handling type-checking - logged in issue #2141 .

MikeSpreitzer commented on pkg/status/combinedstatus-resolution.go:
```diff
@@ -0,0 +1,308 @@
+// associated statuscollector has not been processed.
+// The function returns a tuple (bool1, bool2):
+//
+// - bool1: true if one or more statuscollectors were removed.
+//
+// - bool2: true if one or more statuscollectors were added.
```
Maybe give them more meaningful names, like "removedSome" and "addedSome"?

MikeSpreitzer commented on pkg/status/combinedstatus-resolution.go:
```diff
@@ -0,0 +1,308 @@
+		}
+	}
+
+	// add new statuscollector data and queue up all workstatuses. this does not qualify as
```
It literally _is_ a change in the `combinedStatusResolution`. What sort of change does this not qualify as?

vMaroon replied:
This comment was left over. Overall, the code went a design update cycle not long ago, and I published before reviewing.

MikeSpreitzer commented on pkg/status/combinedstatus-resolution.go:
```diff
@@ -0,0 +1,308 @@
+
+// setStatusCollectors sets ALL the statuscollectors relevant to the
+// combinedstatus resolution.
+// The given map is expected to be thread-safe, its values contain valid
```
A primitive map is not thread-safe. I think you mean to say that nothing mutates the map during a call on this method.

MikeSpreitzer commented on pkg/status/combinedstatus-resolver.go:
```diff
@@ -0,0 +1,307 @@
+	// compare contents: all names of expressions must common, and their expressions must be equal
+	// select-clause first since groupBy and combinedFields would be empty if select is not
+	selectClauseMap := abstract.SliceToMap(scc.selectClause,
+		func(ne v1alpha1.NamedExpression) string { return ne.Name },
+		func(ne v1alpha1.NamedExpression) v1alpha1.Expression { return ne.Def })
```
Write these functions once at file scope rather than repeatedly inside this method.

MikeSpreitzer commented on pkg/status/combinedstatus-resolution.go:
```diff
@@ -0,0 +1,308 @@
+			continue
+		}
+
+		if !statusCollectorData.equals(clauses) {
```
This reads oddly, it is asking whether a statusCollectorData _equals_ a clauses. But you mean to ask whether the _clauses of_ the statusCollectorData equals another clauses. It would be better if this were more obvious here and at the other place like it.

MikeSpreitzer commented on pkg/status/combinedstatus-resolution.go:
```diff
@@ -0,0 +1,308 @@
+	logger := klog.FromContext(ctx)
+
+	updated := false
+	for statusCollectorName := range c.statusCollectorNameToData {
+		scData, ok := c.statusCollectorNameToData[statusCollectorName]
+		if !ok {
+			continue
+		}
```
Why not just

```
for statusCollectorName, scData := range c.statusCollectorNameToData {
```

?

vMaroon replied:
I have to say I have no recollection of how this piece came to life. Again the code went through a change in implementation design, and I though publishing it before finalizing self-reviews to give the general idea (through the comments, mostly) would be beneficial. 

Though the latter is true regardless, I just apologize for having you check these instead.

MikeSpreitzer replied:
No worries. I understand that this is a work in progress. Appreciate the early view.

MikeSpreitzer commented on pkg/status/combinedstatus-resolution.go:
```diff
@@ -0,0 +1,308 @@
+// evaluateWorkStatusAgainstStatusCollector evaluates the workstatus against
+// the statuscollector clauses and caches the evaluations. The function returns
+// true if an evaluation is updated.
+// If any evaluation fails, the function returns an error.
```
The API design is a little more structured.

https://github.com/kubestellar/kubestellar/blob/v0.23.0-alpha.3/api/control/v1alpha1/types.go#L419-L420

https://github.com/kubestellar/kubestellar/blob/v0.23.0-alpha.3/api/control/v1alpha1/types.go#L353 . I will admit that I was assuming when I wrote that that an SQL-like treatment of "undefined" values and that it would be conflated with "null". I do not think that Eval errors should abort processing; I think that they should have defined effects on the fields of a StatusCollectorSpec. As I tried to say in this case, an undefined value (AKA Eval error) should be treated as failing to pass the filter.

Eval error in GroupBy should map to "the undefined value".
Same for Select.
For aggregations, Eval error in the NamedAggregator.Subject should exclude that row (WorkStatus) from the aggregation's input.

vMaroon replied:
I was under the impression that we'd want to abort the whole statement if part of it is bad, similarly to how most SQL dialects behave if im not mistaken.

MikeSpreitzer replied:
It is a design question worth discussing.
I was thinking that in a large dynamic system, there will be the occasional thing that is unexpected. I did not want that to ruin the aggregation of stuff with the expected shape. I also worried about the volume of complaints if every exception is reported.

vMaroon replied:
I think it's sensible to log these errors and not propagate them up the stack for now. Just to be clear, the failure of evaluating one workstatus against a single statuscollector should not affect other statuscollectors in the combinedstatus. 

It is highly likely that if one workstatus fails a statuscollector clause then all other associated workstatuses would as well. In the cases where this does not hold, aggregated status would still be calculated using the rest. I believe it's simpler to abort in this code iteration - what do you think?

vMaroon replied:
Fixed

MikeSpreitzer replied:
I think that the user should get some state-based reporting of evaluation errors.

Here is a proposal. The first question to answer is: how to the errors appear in the API? Let us divide the evaluation errors into two categories: (1) those that arise while creating the materialized views, and (2) those that arise while reducing those to the NamedStatusCombination structs. Note that these categories of error arise at different times: (1) when updating a materialized view, (2) when reducing such a view by aggregation (creating/updating a CombindStatus object). For category 1, perhaps we could extend NamedStatusCombination with the ability to report (a) one error of this category (including a display of the error and the name of the WEC that it comes from --- and insist on taking the error from the lexicographically first WEC with such an error, for the sake of determinism) and (b) a count of the number of rows in the materialized view that are missing due to evaluation errors. For category 2, perhaps we could also extend NamedStatusCombination with the ability to report one error per column, reporting the problem in terms of values without attempting to identify which row of the result would exist if the error had not arisen.

For category 1 errors, the implementation would extend workStatusData with the ability to hold one `error` of category 1.  The one that gets recorded is the first one that arises when trying to do the evaluations for that workStatusData. For category 2 errors, the data structure does not need to be extended.

vMaroon replied:
Will be addressed in followup PR - logged in issue #2141.

MikeSpreitzer commented on pkg/status/combinedstatus-resolution.go:
```diff
@@ -0,0 +1,308 @@
+	// evaluate select
+	selectEvals := make(map[string]any)
+	for _, selectNamedExp := range scData.selectClause {
+		eval, err := celEvaluator.Evaluate(string(selectNamedExp.Def), workStatusContent)
```
Maybe redefine `celEvaluator.Evaluate` so that every call site does not need a cast?

MikeSpreitzer commented on pkg/status/combinedstatus-resolution.go:
```diff
@@ -0,0 +1,308 @@
+	}
+
+	// update the workstatus data
+	wsData.selectEval = selectEvals
+	wsData.groupByEval = groupByEvals
+	wsData.combinedFieldsEval = combinedFieldEvals
```
`wsData` is not a pointer.

MikeSpreitzer replied:
.. so these asignments to a local struct that is not returned have no meaningful effect.

vMaroon replied:
Are you referring to the maintenance of CombinedStatus objects? Such as when changed externally or deleted for whatever reason? This will be handled when the TODO stated in the resolver (`// TODO: API for generating/comparing combinedstatus resolutions`) is addressed and used.

MikeSpreitzer commented on pkg/status/combinedstatus-resolver.go:
```diff
@@ -0,0 +1,307 @@
+	// In the case of creation, the binding.Resolution's OwnerReference is used
+	// to identify the bindingpolicy that the resolution is associated with.
+	//
+	// The returned array contains:
```
It's not an array.

Since it contains two sorts of things, maybe return two homogenous sets?

MikeSpreitzer commented on pkg/status/combinedstatus-resolver.go:
```diff
@@ -0,0 +1,307 @@
+	"github.com/kubestellar/kubestellar/pkg/util"
+)
+
+type CombinedStatusResolver interface {
```
I had envisioned that an expression can refer to both desired state and reported state of a workload object. To support that, this interface needs a method to be told about the content or deletion of the desired state for a particular workload object.

vMaroon replied:
Are you referring to the maintenance of CombinedStatus objects? Such as when changed externally or deleted for whatever reason? This will be handled when the TODO stated in the resolver (// TODO: API for generating/comparing combinedstatus resolutions) is addressed and used.

MikeSpreitzer replied:
No, I am referring to being able to do stuff like write an expression that compares fields from `spec` and `status` (e.g., `status.availableReplicas != spec.replicas`.

vMaroon replied:
Will be addressed in followup PR - logged in issue #2141.

MikeSpreitzer commented on pkg/status/combinedstatus-resolution.go:
```diff
@@ -0,0 +1,308 @@
+	statusCollectorNameToData map[string]*statusCollectorData
+	// workStatusToStatusCollectorName is a map of workstatus references to
+	// a set of statuscollector names that they contribute to.
+	workStatusToStatusCollectorName map[util.ObjectIdentifier]sets.Set[string]
```
I do not see why this field is needed.

vMaroon replied:
Also a left-over.

MikeSpreitzer commented:
I think that the overall strategy for this implementation should be documented somewhere. I had to read too much detailed code to get the big picture.

I think that the big picture could be simpler and better regarding what gets queued and worked on asynchronously. Remember that the reason that a controller does not do all its work in the informer event handlers is that they need to complete quickly. Since a binding.Resolution is already being updated in a worker goroutine, its event handlers do _not_ have such urgency. The design would be simpler if the status combination data structure were more thoroughly updated in the binding.Resolution event handlers (callbacks). I think that there _is_ one good use for queuing, which is to batch re-evaluation of the aggregations. With this change, the big picture would look like the following.

The status controller maintains materialized views of the rows to be aggregated. Any update to such a materialized view causes a call to `DelayingInterface.AddAfter` with the ID of a `CombinedStatus` object to be re-evaluated at a later time. Using a delay means that quickly-spaced materialized view updates cause only one re-evaluation of the aggregation. Each materialized view is specific to the spec of one StatusCollector, and re-created whenever that spec changes. In brief, the data structures involved look like the following.

CollectorSpecs: collector_name -> StatusCollectorSpec

CSRs: binding_id -> wload_obj_id -> CSR

CSR: collector_name -> SCData

SCData: wec_id -> EvaldRow // but for rows that pass the filter

SCData holds the materialized view for one (CombinedStatus, StatusCollector) pair.

EvaldRow is another name for what the current PR draft calls `workStatusData`, except no need for the `filterEval bool` field because it can only be true in a row that actually appears in the data structure.

Any change to a relevant input causes immediate updating of this data structure, yielding a set of CombinedStatus object IDs to enqueue for eventual re-evaluation. Evaluating such a CombinedStatus object ID is done by iterating over the relevant StatusCollector objects and, for each one, evaluating the aggregation of the corresponding materialized view. Once that is done for all the relevant StatusCollectors, the new value of the CombinedStatus object is written out.

MikeSpreitzer commented:
Maintaining materialized views saves computation at the expense of memory. It is going to be very profligate in memory usage. Ultimately we have to shard anyway. It may be better to spend less on memory and more on computation. Following is an outline that spends less memory.

In this design, the status combination data structure does not maintain the results of Expression evaluation. Note that the informer on WorkStatus objects is already holding the latest rev of each in memory. The status combination data structure holds pointers to those objects, indexed by the needs of status combination. Upon a relevant change in inputs, a reference to a CombinedStatus object is enqueued (with `DelayingInterface.AddAfter`) for eventual re-evaluation. Such evaluation applies the relevant StatusCollector specs to the indexed WorkStatus objects. In brief, the data structure looks like the following.

CollectorSpecs: collector_name -> StatusCollectorSpec

CSIs: binding_id -> wload_obj_id -> CSI

CSI: struct {
    reported: wec_id -> `*WorkStatus`
}

MikeSpreitzer commented:
Both of the above designs continue to eschew the designed type checking. To recover that would involve introducing an intermediate layer of data structure per GVR and also per (GVR, StatusCollector). I think this would amount to replacing those `foo -> wload_obj_id -> bar` with `foo -> GVR -> rsc_data; type rsc_data = struct {bars map[NamespacedName]bar; collectors map[string]digestedCollector}` and holding the results of type checking in digestedCollector.

vMaroon commented:
@MikeSpreitzer thank you for reviewing, I have partially addressed comments (comments on CEL later).
On your proposals, I think the first is pretty close to what is intended. The reason I chose to spend memory rather than computation is because I assumed that triggering a fleet of evaluations per individual change is costly.

I think we can later optimize the management of stored information, I prefer this over the alternative.

MikeSpreitzer commented on pkg/status/combinedstatus-resolver.go:
```diff
@@ -0,0 +1,290 @@
+	// 3. For every workload object associated with one or more statuscollectors, a
+	// combinedstatus resolution is created/updated.
+	//
+	// The returned value consists of two homogenous sets
```
The signature says that only one Set is returned, and the comment says that it is not homogeneous.

MikeSpreitzer commented on pkg/status/combinedstatus-resolution.go:
```diff
@@ -0,0 +1,278 @@
+// does not match the filter, no other clause is evaluated.
+// The function returns true if an evaluation is updated.
+// If any evaluation fails, the function returns an error.
+// The function assumes that the caller holds a lock over the combinedstatus
```
I usually use a naming convention to make locking discipline easier to check. The name of a method that is called with the lock held ends with "Locked".

MikeSpreitzer commented on pkg/status/combinedstatus-resolution.go:
```diff
@@ -0,0 +1,278 @@
+		}
+
+		if !eval.Value().(bool) && !exists { // workstatus not previously evaluated and still irrelevant
+			return false, nil
```
What about undoing the `scData.workStatusRefToData[workStatusWECName] = wsData` that was done above?

MikeSpreitzer commented on pkg/status/combinedstatus-resolution.go:
```diff
@@ -0,0 +1,278 @@
+
+		if !eval.Value().(bool) && !exists { // workstatus not previously evaluated and still irrelevant
+			return false, nil
+		}
```
Shouldn't this also remote the entry from `scData.workStatusRefToData` if `!eval.Value().(bool) && exists`?

MikeSpreitzer commented on pkg/status/combinedstatus-resolver.go:
```diff
@@ -0,0 +1,295 @@
+	// This function returns:
+	//   - The identifiers of combinedstatus objects that should be queued for
+	//     syncing.
+	//   - The identifiers of workload objects that should have their
+	//     associated workstatuses processed for evaluation.
+	//   - The names of statuscollectors that should be fetched.
```
Part of what I was suggesting in https://github.com/kubestellar/kubestellar/pull/2174#issuecomment-2138773364 is to not defer these kinds of work but rather for NoteBindingResolution to immediately do it. There will be informers on StatusCollector and WorkStatus (and eventually workload objects), right? An informer maintains a local cache of its objects.

vMaroon replied:
I am aware of this possibility. I think it's hard to tell whether this or what I proposed is better, but both are fine by me - shifting to this approach. My thinking was to have each workstatus fully take care of itself and everything that uses it.

MikeSpreitzer commented on pkg/status/combinedstatus-resolution.go:
```diff
@@ -0,0 +1,333 @@
+	return updated, nil
+}
+
+// evaluateWorkStatusAgainstStatusCollector evaluates the workstatus against
```
Name update here too.

MikeSpreitzer commented on pkg/status/combinedstatus-resolution.go:
```diff
@@ -0,0 +1,333 @@
+
+	if !exists {
+		wsData = &workStatusData{}
+		scData.workStatusRefToData[workStatusWECName] = wsData
```
Set `updated` to `true` in this case?

MikeSpreitzer commented on pkg/status/combinedstatus-resolution.go:
```diff
@@ -0,0 +1,333 @@
+			return false, fmt.Errorf("failed to evaluate select expression: %w", err)
+		}
+
+		updated = updated || eval.Value() != wsData.selectEval[selectNamedExp.Name]
```
This equality test looks wrong. `eval` and the RHS are of type `ref.Val`. Isn't the right way to test using the `Equal` method of `ref.Val`? Invoking `.Value()` on one side and not the other seems suspicious to me.

MikeSpreitzer commented on pkg/status/combinedstatus-resolver.go:
```diff
@@ -0,0 +1,470 @@
+	// The update may involve adding or removing statuscollectors, and changing
+	// the set of destinations associated with the binding.
+	//
+	// The function uses the statuscollector lister and the cached workstatus
+	// objects to update internal state. The given array is assumed to contain
+	// all objects that may be relevant to the binding.
```
What is meant here by "the cached workstatus objects"? I am not sure what cache this is talking about. I see another sentence that I think is about the given array (really slice) of `workStatus` objects, that one seems fairly clear.

vMaroon replied:
The intention here is that these objects should be brought from the informer's cache, ideally only the ones relevant to the calculation. Changed.

MikeSpreitzer commented on pkg/status/combinedstatus-resolver.go:
```diff
@@ -0,0 +1,470 @@
+
+type combinedStatusResolver struct {
+	sync.RWMutex
+	celEvaluator *celEvaluator
```
I think that it is conventional to list immutable and thread-safe fields before the mutex.

I am having trouble finding documentation asserting this, but I think that `cel-go/cel.Env` is thread-safe.

vMaroon replied:
Thanks

MikeSpreitzer commented on pkg/status/combinedstatus-resolver.go:
```diff
@@ -0,0 +1,470 @@
+	// GenerateCombinedStatus generates a CombinedStatus object for the given
+	// binding name and workload object identifier.
+	// If no resolution is associated with the given combination, nil is returned.
+	GenerateCombinedStatus(bindingName string, objectIdentifier util.ObjectIdentifier) *v1alpha1.CombinedStatus
```
Remember that the name of the `CombinedStatus` is composed from the UIDs of the BindingPolicy (oops, probably should be the Binding) object and the workload object.

vMaroon replied:
@MikeSpreitzer fixed

MikeSpreitzer commented on pkg/status/combinedstatus-resolver.go:
```diff
@@ -0,0 +1,470 @@
+			runtime2.HandleError(fmt.Errorf("failed to get statuscollector %s: %w", statusCollectorName, err))
+		}
+
+		c.statusCollectorNameToSpec[statusCollectorName] = &statusCollector.Spec // readonly
```
nil pointer fault if `err != nil` ?

MikeSpreitzer commented on pkg/status/combinedstatus-resolver.go:
```diff
@@ -0,0 +1,470 @@
+	}
+
+	// if the binding resolution is not yet noted - create a new entry
+	objectIdentifierToResolutions, exists := c.bindingNameToResolutions[bindingName]
```
FYI, I would name it `objectIdentifierToResolution` (singular) because each OID is mapped to just one resolution.

MikeSpreitzer commented on pkg/status/combinedstatus-resolver.go:
```diff
@@ -0,0 +1,470 @@
+		return false
+	}
+
+	if spec1.Filter != spec2.Filter {
```
This is not the right way to compare pointers to strings.

MikeSpreitzer replied:
I think that this func needs to compare the referenced strings, not the pointers to the strings.

MikeSpreitzer commented on pkg/status/combinedstatus-resolution.go:
```diff
@@ -0,0 +1,333 @@
+			continue
+		}
+
+		if statusCollectorSpecsMatch(statusCollectorData.StatusCollectorSpec, statusCollectorSpec) {
```
I think you forgot a negation in the condition here.

MikeSpreitzer commented on pkg/status/combinedstatus-resolver.go:
```diff
@@ -0,0 +1,470 @@
+// with the two are skipped.
+// The returned set contains the identifiers of combinedstatus objects that
+// should be queued for syncing.
+// The method is expected to be called with the read lock held.
```
Perhaps the naming convention should be to end with "ReadLocked" or "WriteLocked".

MikeSpreitzer commented on pkg/status/combinedstatus.go:
```diff
@@ -0,0 +1,29 @@
+	// - the UID of the workload object
+	// - the string ":"
+	// - the UID of the BindingPolicy object.
+	return util.IdentifierForCombinedStatus("", objectIdentifier.ObjectName.Namespace) // TODO
```
The job is impossible with the given arguments.

vMaroon replied:
That is true. This is a placeholder that will be updated. It will introduce changes one way or another but until it is written - keeping it like this.

vMaroon replied:
@MikeSpreitzer fixed

MikeSpreitzer commented on pkg/status/combinedstatus-resolution.go:
```diff
@@ -0,0 +1,333 @@
+	combinedFieldEvals := make(map[string]ref.Val)
+	for _, combinedFieldNamedAgg := range scData.CombinedFields {
+		if combinedFieldNamedAgg.Type == v1alpha1.AggregatorTypeCount {
+			// count does not require a subject - mark the evaluation with the type
```
The second half of this comment looks outdated.

MikeSpreitzer commented on pkg/status/combinedstatus-resolution.go:
```diff
@@ -0,0 +1,333 @@
+			currentEval, exists := wsData.combinedFieldsEval[combinedFieldNamedAgg.Name]
+			updated = updated || !exists || currentEval != nil
+
+			combinedFieldEvals[combinedFieldNamedAgg.Name] = nil
```
Is it even necessary to have an entry in this map in this case?

vMaroon replied:
If the name exists and it is a nil (and not the case of nil because the key is not mapped) then the type is COUNT. This is needed in order to properly update `updated` upon this kind of change.

MikeSpreitzer replied:
I am not sure that this is the only way to do it. Not a blocker.

MikeSpreitzer commented on pkg/status/combinedstatus-resolution.go:
```diff
@@ -0,0 +1,333 @@
+			continue
+		}
+
+		// evaluate subject which should not be nil since the statuscollector is valid
```
But I do not think that there is validation that _guarantees_ that Subject is non-nil in this case.

vMaroon replied:
This comment refers to Subject in the CombinedFields of the StatusCollector spec. The function assumes validity of the status collector (which is not yet implemented), and given that the block above checks the COUNT case, at this stage a Subject must be present in spec.

This does not mean that evaluating subject guarantees a non-nil.

MikeSpreitzer replied:
> The function assumes validity of the status collector (which is not yet implemented)

Yes, that is my concern. This statement will be buggy until that validation is added or this statement is changed to not assume that absent validation.

vMaroon replied:
Will be addressed in followup PR - logged in issue #2141.

MikeSpreitzer commented on pkg/status/combinedstatus-resolution.go:
```diff
@@ -0,0 +1,333 @@
+			return false, fmt.Errorf("failed to evaluate combinedFields expression: %w", err)
+		}
+
+		updated = updated || eval.Value() != wsData.combinedFieldsEval[combinedFieldNamedAgg.Name]
```
This equality test is also suspect.

MikeSpreitzer commented on pkg/status/combinedstatus-resolver.go:
```diff
@@ -0,0 +1,470 @@
+	// The returned set contains the identifiers of combinedstatus objects
+	// that should be queued for syncing.
+	NoteBindingResolution(bindingName string, bindingResolution binding.Resolution, deleted bool,
+		workStatuses []*workStatus,
```
There could be a lot of these that are irrelevant, right? Wouldn't it be more efficient to pass a function that returns the needed WorkStatus (which would be implemented by a fetch from the local cache of the WorkStatus informer)?

vMaroon replied:
Whoever calls NoteBindingResolution has enough information to limit the passed list to what is necessary. Unfortunately, at the moment, we can only filter fetching from the informer's cache by WDS name. WorkStatuses are labelled with the name of a BindingPolicy but I'm not sure how that addresses multiple BPs selecting the same objects.

Therefore, the fetching is at a higher level, and when consuming workstatuses there is filtering to ignore what's irrelevant.

MikeSpreitzer replied:
Remember that when I talk about an "informer" I mean a `SharedIndexInformer`. These thing support indexing. This controller could add an index that supports fetching the exact one WorkStatus that is needed at a given point in the code.

vMaroon replied:
Check this https://github.com/vMaroon/kubestellar/pull/11/files#r1642771147

MikeSpreitzer commented on pkg/status/combinedstatus-resolution.go:
```diff
@@ -0,0 +1,335 @@
+			return false, fmt.Errorf("failed to evaluate select expression: %w", err)
+		}
+
+		updated = updated || eval.Value() != wsData.selectEval[selectNamedExp.Name].Value()
```
Wouldn't it be safer and potentially more efficient to use the `Equal` method of `ref.Val`?

I see that the restriction that such an Expression refers to an atomic piece of data is no longer there. I am not sure whether that is deliberate. It does seem better to me if we do not have this restriction. The Go language `==` on slices and maps is surely not what we would want in such a case.

MikeSpreitzer commented on pkg/status/combinedstatus-resolver.go:
```diff
@@ -0,0 +1,472 @@
+			abstract.SliceToPrimitiveMap(objectData.StatusCollectors,
+				func(statusCollectorName string) string { return statusCollectorName }, // keys mapper
+				func(statusCollectorName string) *v1alpha1.StatusCollectorSpec { // val mapper
+					return c.statusCollectorNameToSpec[statusCollectorName] // if missing after fetch, ignored
```
More precisely, the value returned here will be `nil` if the earlier fetch failed on this statusCollectorName. I suspect that `csResolution.setStatusCollectors` will do the wrong thing for the `map` entry created in this case.

vMaroon replied:
`csResolution.setStatusCollectors` would create an entry for the status collector with a nil spec. While this would not do anything harmful (as it would be expected for ::updateStatusCollector to be called with the same name at a certain stage) - it would mark the combined-status resolution as updated, which is inaccurate. Fixing the latter.

MikeSpreitzer commented on pkg/status/combinedstatus-resolver.go:
```diff
@@ -0,0 +1,472 @@
+		}
+
+		// fetch missing statuscollector specs
+		c.fetchMissingStatusCollectorSpecsLocked(statusCollectorLister, objectData.StatusCollectors)
```
FYI, the method naming convention should be explained up front somewhere.

vMaroon replied:
Will be addressed in followup PR - logged in issue #2141.

andreyod commented on pkg/status/combinedstatus-resolver.go:
```diff
@@ -0,0 +1,472 @@
+
+	return &combinedStatusResolver{
+		celEvaluator:             celEvaluator,
+		bindingNameToResolutions: make(map[string]map[util.ObjectIdentifier]*combinedStatusResolution),
```
Initialize also the `statusCollectorNameToSpec` map here.

andreyod commented on pkg/status/workstatus.go:
```diff
@@ -0,0 +1,29 @@
+type workStatus struct {
+	wecName                string
+	sourceObjectIdentifier util.ObjectIdentifier
+	status                 *runtime.RawExtension
```
Maybe this field type should be `map[string]interface{}`? This will prevent unnecessary conversion.

vMaroon replied:
I prefer keeping this struct identical to the workstatus. Overall, someone will have to do this conversion since this is what WorkStatus objects contain.

andreyod replied:
The caller code gets the `map[string]interface{}` here https://github.com/vMaroon/kubestellar/pull/11/files#diff-b454b6b87e03b38df5fd12bcd7217e0dd525d4278d5b02c5c4de2c5e2a1f83c4R85, and the consuming code converts to `map[string]interface{}` here https://github.com/kubestellar/kubestellar/pull/2174/files#diff-faa0ed061f946759218486d1b63d7652398765a0e05099b16501ce467c6285d5R89
So if we change the `workStatus.status` type to `map[string]interface{}` there will be no conversion at all.

vMaroon replied:
Ah sorry I did not see that yet, will review (and change this field)

MikeSpreitzer commented on pkg/status/combinedstatus-resolution.go:
```diff
@@ -0,0 +1,358 @@
+			continue
+		}
+
+		if !statusCollectorSpecsMatch(statusCollectorData.StatusCollectorSpec, statusCollectorSpec) {
```
nil pointer fault inside statusCollectorSpecsMatch when `statusCollectorSpecsMatch == nil`.

MikeSpreitzer commented on pkg/status/combinedstatus-resolution.go:
```diff
@@ -0,0 +1,358 @@
+	// add new statuscollector data
+	for statusCollectorName, statusCollectorSpec := range statusCollectorNameToSpec {
+		if statusCollectorSpec == nil {
+			continue // not cached
```
I think the code would be easier to understand if a `name: nil` entry in `statusCollectorNameToSpec` were given the same semantics as `name` not being in the map. That would mean the data structure has a state-based consistency predicate. Taking `name:nil` to mean "no change" makes the proper data structure contents depend on more history than just the latest information.

vMaroon replied:
done.

MikeSpreitzer commented on pkg/status/combinedstatus-resolver.go:
```diff
@@ -0,0 +1,520 @@
+	NoteStatusCollector(statusCollector *v1alpha1.StatusCollector, deleted bool,
+		workStatusLister cache.GenericLister) sets.Set[util.ObjectIdentifier]
+
+	// NoteWorkStatus notes a workstatus in the combinedstatus resolution
```
There may be more than one associated CombinedStatus (e.g., when there are multiple Binding objects that reference the same workload object).

vMaroon replied:
typo, fixed

MikeSpreitzer commented on pkg/status/combinedstatus-resolver.go:
```diff
@@ -0,0 +1,520 @@
+
+// NewCombinedStatusResolver creates a new CombinedStatusResolver.
+func NewCombinedStatusResolver(celEvaluator *celEvaluator) (CombinedStatusResolver, error) {
+	celEvaluator, err := newCELEvaluator()
```
Why not use the given celEvaluator?

MikeSpreitzer commented on pkg/status/combinedstatus-resolver.go:
```diff
@@ -0,0 +1,520 @@
+	// if the binding resolution is not yet noted - create a new entry
+	objectIdentifierToResolution, exists := c.bindingNameToResolutions[bindingName]
+	if !exists {
+		objectIdentifierToResolution = make(map[util.ObjectIdentifier]*combinedStatusResolution)
```
BTW, the desired size of this map is known at this point.
Not a blocker.

MikeSpreitzer commented on pkg/status/combinedstatus-resolution.go:
```diff
@@ -0,0 +1,358 @@
+	statusCollectorNameToData map[string]*statusCollectorData
+	// collectionDestinations is a set of destinations that are expected to be
+	// collected from.
+	collectionDestinations sets.Set[string]
```
Why does there need to be a field holding destinations that are _expected to be_ collected from?
`statusCollectorNameToData` holds, in every `statusCollectorData`, the names of the WECs that _have been_ collected from, and resolver.NoteBindingResolution carries the new set of destinations on the stack.

vMaroon replied:
I replaced this with making `combinedStatusResolution::evaluateWorkStatus` requires a boolean that determines that a workstatus should be added if missing or not.

vMaroon replied:
`statusCollectorData` holds the names of the WECs that passed the filtering. Therefore, the required knowledge has to be maintained here. In a previous iteration I did map such WEC names to nils, but afterwards this set was added.

MikeSpreitzer replied:
`statusCollectorNameToData` holds, in its nested `statusCollectorData` objects, the names of the WECs that passed the filtering.

When updating the data structure in reaction to a new binding resolution, resolver.NoteBindingResolution carries in its stack the new set of names of WECs. When updating for a change to anything else, there is no change in the set of names of the WECs so the representation in `statusCollectorNameToData..statusCollectorData` is adequate.

vMaroon replied:
In the case of a statuscollector update, it is possible that the filter changes which may lead to the set of passing workstatuses to change.

MikeSpreitzer replied:
Ah, right.

MikeSpreitzer commented on pkg/status/combinedstatus-resolver.go:
```diff
@@ -0,0 +1,520 @@
+		// should evaluate workstatuses if added/updated collectors or added destinations
+		// it's possible to evaluate workloads coming from new destinations only if no collectors were removed (TODO)
+		if addedCollectors || len(newDestinationsSet) > 0 {
+			workloadIdentifiersToEvaluate.Insert(objectIdentifier)
```
The current approach here will do more work than necessary. To explain, let me define the following symbols. All of the remarks here are scoped to one workload object (one run through the body of the loop over `bindingResolution.ObjectIdentifierToData`).

- `collectorsToRemove`, the ones that were removed in `csResolution.setStatusCollectors`
- `collectorsToKeep`, the ones that were already in the domain of `csResolution.statusCollectorNameToData` at the start of this func AND are listed in `objectData.StatusCollectors`.
- `collectorsToAdd`, the ones listed in `objectData.StatusCollectors` BUT NOT already in the domain of  `csResolution.statusCollectorNameToData` at the start of this func.

- `destinationsToRemove`, the ones that had data in `csResolution` at the start of this func but are NOT listed in `bindingResolution.Destinations`.
- `destinationsToKeep`, the ones that had data in `csResolution` at the start of this func AND ARE listed in `bindingResolution.Destinations.
- `destinationsToAdd`, the ones that are in `bindingResolution.Destinations` BUT did NOT have data in `csResolution` at the start of this func.

At the start of this func, csResolution had materialized view rows for the cross product of (collectorsToRemove U collectorsToKeep) X (destinationsToRemove U destinationsToKeep) (here I use "U" for set union). This func needs to change csResolution so that it has rows for the cross product of (collectorsToKeep U collectorsToAdd) X (destinationsToKeep U destinationsToAdd). At this point in the code, three blocks has already been removed:

- collectorsToRemove X destinationsToRemove,
- collectorsToRemove X destinationsToKeep, and
- collectorsToKeep X desinationsToRemove

The rows in the block collectorsToKeep X destinationsToKeep have been left untouched and should remain. Because this func is responding to a notifcation about a state change in a binding resolution rather than a workload object or a StatusCollector, the rows in collectorsToKeep X destinationsToKeep do not need to be re-evaluated.

What remains to be done is add rows for three blocks:

- collectorsToAdd X destinationsToAdd,
- collectorsToAdd X destinationsToKeep, and
- collectorsToKeep X destinationsToAdd.

These will be new rows, and comparison with previous contents does not need to be done.

vMaroon replied:
Thank you for putting it cleanly. I am aware of the excessive work done here, will indeed be tightened in a follow-up PR.

MikeSpreitzer commented on pkg/status/combinedstatus-resolver.go:
```diff
@@ -0,0 +1,520 @@
+		}
+
+		// evaluate workstatuses
+		for _, workStatus := range workStatuses {
+			// skip workstatuses that are not associated with the binding's destinations
+			if destinations.Has(workStatus.wecName) {
+				continue
+			}
```
Rather than fetch and convert all of the WorkStatus objects and then throw away the ones wanted, this code could iterate over `destinations` and for each one just fetch and convert the one WorkStatus object for that destination.

This is an efficiency issue, not correctness. Can be addressed in follow-on.

vMaroon replied:
thanks

MikeSpreitzer commented on pkg/status/combinedstatus-resolver.go:
```diff
@@ -0,0 +1,500 @@
+
+			objs, err := workStatusIndexer.ByIndex(workStatusIdentificationIndexKey, indexKey) // one obj expected
+			if err != nil {
+				workloadObjectToError[workloadObjIdentifier] = err
```
I do not think user behavior can cause an error to appear here, only bugs can do that.

MikeSpreitzer commented on pkg/status/combinedstatus-resolver.go:
```diff
@@ -0,0 +1,500 @@
+			}
+
+			if len(objs) == 0 {
+				continue
```
Another "no change". Shouldn't csResolution be deleted here?

vMaroon replied:
This condition checks whether a single expected workstatus exists or not. In the case that no expected workstatus was fetched at all, I still don't believe we should delete the entry here. We can instead have the reconciliation of an empty combinedstatus resolution delete the entry. What do you think?

MikeSpreitzer replied:
When the line I am commenting on is executed, are we guaranteed that the `csResolution` is empty? It could have been created and populated with content earlier, right?

vMaroon replied:
We are not guaranteed it is empty. Yes. In both cases that this whole call leads to a populated or empty combinedstatus, the combinedstatus identifier would be added to the set of identifiers to queue higher on the stack.

When syncing a combinedstatus (pending), an entry should be deleted if empty. This is what I meant above.

MikeSpreitzer commented on pkg/status/controller.go:
```diff
@@ -140,6 +286,18 @@ func (c *Controller) runWorkStatusInformer(ctx context.Context) {
+			wecName := obj.(metav1.Object).GetNamespace()
+			sourceRef, err := util.GetWorkStatusSourceRef(obj.(runtime.Object))
+			if err != nil {
+				return nil, err
```
Fun fact: the informer library code will panic if this func returns a non-nil error.

MikeSpreitzer commented on pkg/status/combinedstatus-resolver.go:
```diff
@@ -0,0 +1,500 @@
+
+			workStatus, err := convertToWorkStatus(objs[0].(runtime.Object))
+			if err != nil {
+				workloadObjectToError[workloadObjIdentifier] = err
```
Also not a user error, right?

MikeSpreitzer commented on pkg/binding/bindingpolicy-resolution-broker.go:
```diff
@@ -101,7 +111,7 @@ func (broker *resolutionBroker) GetResolution(bindingPolicyKey string) *Resoluti
-		ObjectIdentifierToData: abstract.PrimitiveMapSafeMap(&bindingPolicyResolution.RWMutex,
+		ObjectIdentifierToData: abstract.PrimitiveMapSafeValMap(&bindingPolicyResolution.RWMutex,
 			bindingPolicyResolution.objectIdentifierToData,
 			func(data *objectData) struct {
 				ResourceVersion  string
```
Please establish a type definition for this struct type.

MikeSpreitzer commented on pkg/status/controller.go:
```diff
@@ -132,6 +201,83 @@ func (c *Controller) run(ctx context.Context, workers int, cListers chan interfa
+}
+
+func (c *Controller) enqueueCombinedStatus(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
```
FYI: `obj` will never be a `cache.DeletedFinalStateUnknown`, so this line could call `cache.MetaNamespaceKeyfunc` or `cache.ObjectToName` instead.

MikeSpreitzer commented on pkg/status/controller.go:
```diff
@@ -132,6 +201,83 @@ func (c *Controller) run(ctx context.Context, workers int, cListers chan interfa
+	return nil
+}
+
+func (c *Controller) enqueueCombinedStatus(obj interface{}) {
```
FYI: the type of `obj` could be `metav1.Object`, and the next line could call `cache.MetaObjectToName` without needing any error handling.

MikeSpreitzer commented on pkg/status/controller.go:
```diff
@@ -196,7 +354,7 @@ func (c *Controller) enqueueObject(obj interface{}) {
 		return
 	}
-	c.workqueue.Add(ref)
+	c.workqueue.Add(workStatusRef(ref.String()))
```
FYI: The `handleObject` and `enqueueObject` methods of Controller are poorly named. I had assumed they apply to workload objects. But actually they apply to WorkStatus objects. It would be better if their names said so.

Not a blocker.

MikeSpreitzer commented on pkg/status/workstatus.go:
```diff
@@ -0,0 +1,162 @@
+}
+
+func (c *Controller) syncWorkStatus(ctx context.Context, key string) error {
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
```
FYI: If instead you had defined `type workStatusRef cache.ObjectName` then there would be no need to call a parsing function here and no need for error handling here.

Not a blocker.

vMaroon replied:
The current WorkStatus name is not safe to work with for inferring source object / wecname in my opinion. Therefore I changed this ref to the portion of the `workStatus` struct that deals with bookkeeping.

MikeSpreitzer replied:
@waltforme

MikeSpreitzer replied:
Remember that we are doing eventually consistent programming with no timing guarantees in a system of components that can individually go up and down. We should assume no limit (apart from what we explicitly program) on what other goroutines and other processes do between two consecutive memory operations in one goroutine.

How much should this package assume about the behavior of code in the ks/OSA repo? How much should this package assume about clients manipulating an ITS in general?

Do we expect the `.spec.sourceRef` of a WorkStatus to be immutable?

It is not immutable today. The IDL does not say that it is immutable. We do not lock down the authority to make changes: the examples, documentation, and tests make liberal use of super-user credentials.

Even if the `.spec.sourceRef` of a WorkStatus were immutable, that would not stop one client from deleting a WorkStatus and then creating a replacement with the same name but different `.spec.sourceRef` (and the status controller adds no finalizer that would prevent that from happening without interlock with the controller). Informers are only eventually consistent; they do not promise to deliver a delete and a create event in this case; the informer client might only see an update event.

If the `.spec.sourceRef` of a WorkStatus _were_ immutable then I think that we could consider it to be an extended part of the object's identity. If we did, then a Get by name would have to be followed by checking the `.spec.sourceRef` and, on mismatch with the expected SourceRef, acting like the Get returned NotFound.

To properly handle `.spec.sourceRef` of a WorkStatus being mutable, the status controller should keep track of what it last got from each WorkStatus --- so that upon change in `.spec.sourceRef` the data previously associated with that WorkStatus can be cleared out.

vMaroon replied:
Will be addressed in followup PR - logged in issue #2141.

MikeSpreitzer commented on pkg/status/workstatus.go:
```diff
@@ -0,0 +1,162 @@
+
+	ws, err := convertToWorkStatus(obj)
+	if err != nil {
+		return err
```
Is this an error that should cause a retry? Will a retry get a different result?

vMaroon replied:
no longer relevant

MikeSpreitzer commented on pkg/status/workstatus.go:
```diff
@@ -0,0 +1,162 @@
+
+	sourceRef, err := util.GetWorkStatusSourceRef(obj)
+	if err != nil {
+		return err
```
Would this error indicate a transient condition that warrants a retry, or a bug that does not, or a user error that does not?

vMaroon replied:
no longer relevant

MikeSpreitzer commented on pkg/status/workstatus.go:
```diff
@@ -0,0 +1,162 @@
+
+	// only process workstatues with the label for single reported status
+	statusLabelVal, ok := obj.(metav1.Object).GetLabels()[util.BindingPolicyLabelSingletonStatusKey]
+	if !ok {
```
Why is this treated differently than `statusLabelVal == util.BindingPolicyLabelSingletonStatusValueUnset` ?

vMaroon replied:
This portion of the code is taken from the previous singleton handling logic.

MikeSpreitzer replied:
Then I am not going to insists on fixing it in this PR.

MikeSpreitzer commented on pkg/status/workstatus.go:
```diff
@@ -0,0 +1,162 @@
+	status, err := util.GetWorkStatusStatus(obj)
+	if err != nil {
+		// status gets updated after workstatus is created, it's ok to requeue
+		return err
```
The error should be returned to cause retry if it is reporting a transient condition, should not be returned if it is reporting a bug.

MikeSpreitzer replied:
If I understand correctly, this statement can be reached only because of bugs. So no error should be returned.

MikeSpreitzer commented on pkg/status/workstatus.go:
```diff
@@ -0,0 +1,162 @@
+
+	obj, err := getObject(lister, objRef.Namespace, objRef.Name)
+	if err != nil {
+		return err
```
The object might have been deleted very recently, this is not an error.

vMaroon replied:
fixed, along with other similar routes in latest commit.

MikeSpreitzer commented on pkg/status/workstatus.go:
```diff
@@ -0,0 +1,162 @@
+	if err != nil {
+		return err
+	}
+	obj, err := getObject(c.workStatusLister, ns, name)
```
Every WorkStatus object is namespaced. No need to go through getObject.

Not a blocker.

MikeSpreitzer commented on pkg/status/statuscollector.go:
```diff
@@ -0,0 +1,50 @@
+		c.workqueue.AddAfter(combinedStatusRef(combinedStatus.ObjectName.AsNamespacedName().String()), queueingDelay)
+	}
+
+	c.logger.Info("Handled StatusCollector", "key", key)
```
FYI: in plain English this is not wrong, but controller code tends to use the following terminology: informer events are "handled", workqueue items are "synced".

MikeSpreitzer commented on pkg/status/combinedstatus.go:
```diff
@@ -0,0 +1,45 @@
+	if err != nil {
+		return err
+	}
+	c.logger.Info("Handled CombinedStatus", "ns", ns, "name", name)
```
`klog.FromContext(ctx)` should be the logger used in this method.

MikeSpreitzer commented on pkg/status/combinedstatus-resolution.go:
```diff
@@ -0,0 +1,356 @@
+
+	// workStatusRefToData is a map of workstatus-hosting WEC name to the
+	// evaluation of the workstatus against the statuscollector's clauses.
+	workStatusRefToData map[string]*workStatusData
```
This only holds an entry if it passed the collector's filter, right? If so then that should be stated in the comment on the field.

MikeSpreitzer replied:
FYI, I find the name "workStatusRefToData" to not be very clearly indicative of what the keys in the map are. I think that a name like "wecToData" would be clearer.

vMaroon commented:
The latest force-push squashes a few commits that address the recent batch of comments and further development.

MikeSpreitzer commented on pkg/status/statuscollector.go:
```diff
@@ -0,0 +1,50 @@
+		c.workqueue.AddAfter(combinedStatusRef(combinedStatus.ObjectName.AsNamespacedName().String()), queueingDelay)
+	}
+
+	c.logger.Info("Synced StatusCollector", "ref", ref)
```
The logger to use here is `klog.FromContext(ctx)`.

MikeSpreitzer commented on pkg/status/workstatus.go:
```diff
@@ -0,0 +1,178 @@
+	} else {
+		status, err := util.GetWorkStatusStatus(obj)
+		if err != nil {
+			return fmt.Errorf("failed to get status from workstatus (%v): %w", ref, err)
```
If I understand correctly, this statement can be reached only because of bugs, not transient conditions.

MikeSpreitzer commented on pkg/status/workstatus.go:
```diff
@@ -0,0 +1,178 @@
+	}
+
+	// only process workstatues with the label for single reported status
+	statusLabelVal, ok := obj.(metav1.Object).GetLabels()[util.BindingPolicyLabelSingletonStatusKey]
```
nil pointer fault here if c.workStatusLister.ByNamespace(ref.wecName).Get(ref.name) returned a NotFound error?

MikeSpreitzer commented on pkg/status/workstatus.go:
```diff
@@ -0,0 +1,178 @@
+		return nil
+	}
+
+	c.logger.Info("updating singleton status", "objectIdentifier", ref.sourceObjectIdentifier)
```
Use the logger from the context in preference to `c.logger`. That is a general design rule.

MikeSpreitzer commented on pkg/status/workstatus.go:
```diff
@@ -0,0 +1,178 @@
+	obj, err := lister.ByNamespace(objectIdentifier.ObjectName.Namespace).Get(objectIdentifier.ObjectName.Name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			logger.V(5).Info("could not find object", "objectIdentifier", objectIdentifier)
```
(A) A log message should be capitalized like an English sentence.

(B) It would be better if this one were a little more informative, saying that an update is being abandoned because the object is not in the local cache.

MikeSpreitzer commented on pkg/status/workstatus.go:
```diff
@@ -0,0 +1,178 @@
+
+	unstrObj, ok := obj.(*unstructured.Unstructured)
+	if !ok {
+		return fmt.Errorf("object cannot be cast to *unstructured.Unstructured: object: %s", util.RefToRuntimeObj(obj))
```
Is this a transient condition that is likely to go away when retried?

vMaroon replied:
This function will still return this error, but the caller will not retry (https://github.com/kubestellar/kubestellar/pull/2174#discussion_r1646316239)

vMaroon commented:
The latest force-push squashes the last batch of code into one commit.

ezrasilvera commented:
/lgtm
/approve

MikeSpreitzer commented on pkg/status/controller.go:
```diff
@@ -181,36 +361,25 @@ func shouldSkipDelete(_ interface{}) bool {
-func (c *Controller) enqueueObjectDeletion(obj interface{}) {
-	c.workqueue.Add(obj)
+	_, ok := obj.(metav1.Object).GetLabels()[util.BindingPolicyLabelSingletonStatusKey]
+	if !ok {
```
Singleton status return is not exclusive with combined status return.
